src/server/ua_services_session.c

```diff
@@ -72,6 +154,42 @@ Service_ActivateSession(UA_Server *server, UA_SecureChannel *channel,
         return;
     }
 
+    if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGN ||
```

Comment:
The content of the if-clause is very much self-contained.
It can be put in a static function.

Do you want to verify + generate the nonce before checking permissions via access control?